### ISSUE TITLE
Keep indent with tabs

### ISSUE THREAD

**Issue by raphaelgoetter** [htacg/tidy-html5#108]
Hello,

Our CSS styleguide ask us to indent using tabulation (1 tab = 4 spaces), but tidy-HTML5 always force tabs to spaces.

Is it possible to indent with tabs on tidy-HTML5 ?

**Comment by balthisar** [comment]
It's a common enough request. I'll keep this open as a feature request for a future milestone.

**Comment by wraybowling** [comment]
+1

**Comment by geoffmcl** [comment]
@raphaelgoetter @balthisar @wraybowling I certainly agree that maybe there are good reasons for an option to pretty printing using tabs instead of spaces. Why not!

How about one of you, or any others, code this? I will certainly help where I can...

While the master branch is headed for a 5.0.0 release, this could be done in say a with-tabs branch for later merging after testing...

Some quick pointers...

#### Adding an option like --use-tabs, TidyUseTabs

Pick an existing boolean option like say wrap-sections, and on searching for this find the enumerated tidy option, TidyWrapSection is this case.

A global source search for this reveals just 4 places where is is used -

 1. enumeration - include\tidyenum.h(141):  `TidyWrapSection, /**< Wrap within <![ ... ]> section tags */`
 2. configuration - src\config.c(264):  `{ TidyWrapSection, PP, "wrap-sections", BL, yes, ParseBool, BoolPicks },`
 3. help text - src\localize.c(792):  `{TidyWrapSection,...`
 4. use - src\pprint.c(1723):  `Bool wrapSect = cfgBool( doc, TidyWrapSection );`

Note the special doxygen comments in tidyenum.h, which are later used to build tidy documentations, along with the 'help text'... just follow the samples...

This particular option default to 'yes', while we might want --use-tabs to default to 'no'.

#### Changing the ouput

Now, in the pprint.c module - `Bool useTabs = cfgBool( doc, TidyUseTabs );` can be used anywhere needed to determine what indent to add...

The pretty printer uses two levels of output. The string is first added to a buffer, which grows in size as output is added. Only on the flush-to-file service is the accumulated string written out, and the used buffer length zeroed.

So this `useTabs` could be used to either alter what is added to the buffer, indent spaces, or tabs, as the out string is built, or possibly just review and change the buffer from spaces to tabs when flushing the output. The former is probably the better way to go, but the latter offers a sort of quick fix.

Then tidy would have a --use-tabs option ;=))

**Comment by geoffmcl** [comment]
@raphaelgoetter, @balthisar, @wraybowling, @Acecool, @hoehrmann, @Vittorio, and others...

Also see a related, very OLD `Feature Request` on sourceforge - https://sourceforge.net/p/tidy/feature-requests/3/

I have now looked carefully into this, and it would be relatively easy to implement ;=)) In pprint.c the leading indent spaces are only output in 3 places, as -

```
        for ( i = 0; i < spaces; ++i )
            TY_(WriteChar)( ' ', doc->docOut );
```

These 3 instances could be changed to -

```
static unit indent_char = ' ';   /* indent character to be used */
  /* then each instance changed to - */
        for ( i = 0; i < spaces; ++i )
            TY_(WriteChar)( indent_char, doc->docOut );
```

And then the config input could set this `indent_char`...

Or these 3 outputs could be re-directed to a single new service, like -

```
   void TY_(WriteIndent)( StreamOut* out, uint spaces );
```

Anyway that's just a question of style, but I have two (2) question that I would seek comments on :-

 1. What should the new option be called? I suggest `--indent-with-tabs yes`. The default is `no`.
 2. How tab output should treat the current indent length, `--indent-spaces integer`, default is 2.
 
Quest 1: @hoehrmann original 2001 post suggested `indent-char: space|tab|none`. Well this is another possibility, but I do not know what the `none` would mean! Would this just be another way of writing `indent: no`, which is the default, or what? That is why I now suggest the simpler `indent-with-tabs: yes|no`, again default `no`.

Quest 2: number of tabs - The simple idea would be to treat this as the number of tabs to output for each indent level, but this length presently defaults to `2`, so that seems faulty... 

One idea would be to re-set this to `1` on reading `indent-with-tabs: yes`, but this would require a new service just to deal with this.

This is not a bad thing, since the simple 'yes' or 'no' option, could also simply change the `indent_char` from the default `space` to a `tab`, and vice versa, so this extra decode service could also do that...

Another idea would be to take this length as a tab unit length. So then the level 1 indent of 2 spaces, means 1 tab is output; a level 2 indent of 4 spaces, means 2 tabs are output, etc, etc... 

That is the current `uint spaces = cfg( doc, TidyIndentSpaces );` is used as a divisor, `if > zero`, to get the count of the tabs to output... this is what I presently favour...

Would certainly appreciate some input... thanks...

I intend to implement this probably this weekend. Probably not before time given the 14 years or so that @Vittorio mentioned in his recent post, jokingly I hope ;=))

**Comment by CTimmerman** [comment]
I use 1 tab per level, and set my editor to make that 4 spaces wide.

`indent-with-tabs: yes|no` (default `no`) simply changing the spaces to tabs would work for me, as i change `indent-spaces` anyway. The new situation would be that i add `indent-with-tabs: yes` and `indent-spaces: 1` instead of only `indent-spaces: 4`.

**Comment by geoffmcl** [comment]
@CTimmerman thanks for your comment. I tend to agree with you that the value of `indent-spaces: 1` should be used as a simple count of tabs per level of indent. And in implmenting this feature, I now re-set indent-spaces to 1, just in case the user forgets. Of course supsequent indent-spaces options can set it to what the user wants...

But since there are likely to be some debatable points about all this, I have implemented my first cut in a new **branch** `indent-with-tabs`. Only after this is tested will I merge it back to master.

To get this branch you only need to do `$ git checkout indent-with-tabs`. This should switch your local clone to this branch, and you may need to do `$ git pull` to make sure you have the latest in that branch... then cmake and compile as normal...

So far in testing everything looks **very good** in normal html output ;=)) 

As expected tabs will be used as the indent character, increasing for each level. All leading space of tags and text has been stripped, so the indent using tabs is fine. And as you point out most modern text editors allow us to set the tab expansion size in the view, and like you I tend to use 4 for this.

The problem I see at the moment is say in the script tag. In here, if say you have used tabs to indent the script, and if you have not changed the (TidyTabSize) default `--tab-size 8`, then these tabs will be expanded to 8 spaces. But even if you set `--tab-size 0` to disable this expansion, or have indented the script code using spaces, then there are some weird things still happening!

This script code is presently output as `text`, but in this case the original leading spaces (or tabs) have been retained for the output. And there are some special services to count any existing leading space, to make sure the indent is not just added to that...

If this is not done then you get what I call **creaping indent** ;=() 

This is where you are now tidying using a previously tidied output, and if not careful, a new indent will be added to the last, now retained input indent. Thus the 2nd tidied output will not matched the tidied input... things will continue to march to the right each time you run tidy on the previous tidy output... **BAD!**

So some of these services, like `TextStartsWithWhitespace( doc->lexer, node, start, mode );`, and `IncrWS( start, end, indent, ixWS );` need to be carefully revisited. And this is tied up with the bit-wise `mode` flag (PREFORMATTED|COMMENT|ATTRIBVALUE|NOWRAP|CDATA) passed around... some work to do here...

And another thing not yet fully tested is how (TidyWrapLen), `--wrap 68` will be effected. In the SF posts I read some concerns about this...

Anyway, this new **branch** `indent-with-tabs` is a start to implement this **feature** after so many years without it ;=)).

And of course it is **OFF** by default so should only effect those who want to **experiment**. 

Look forward to futher comments, especially any patches or pull requests against this branch.

**Comment by geoffmcl** [comment]
Note added 'wt' to the version to make it clear this is a --indent-with-tabs version...

**Comment by geoffmcl** [comment]
To give more exposure to this **NEW** indent-with-tab: yes option have merged the branch into master.

Now this new option will be present from version 4.9.29 onward...

**Comment by geoffmcl** [comment]
Now have initial support for tab output...

No further comments for more than a few weeks so closing this for now...

Feel free to re-open, or post a new issue.

**Comment by Kristinita** [comment]
**Type: Solution** :star:

For HTML Tidy 5.6.0 or higher use [**`keep-tabs` option**](https://api.html-tidy.org/tidy/quickref_next.html#keep-tabs). It:

1. Preserves existing tabs.
1. Doesn’t add unwanted spaces for indentation with tabs.

Thanks.
